fix: `import_utils fetch_archive_from_http` -  improve url parsing for fetching archive from http

malte-aws opened the pull request:
### Related Issues

- fixes #5198

### Proposed Changes:

The extension of the archive was determined by splitting the URL at `"."` using `rpartition` and then using the string to the right of the `"."` as the file extension.

This pull request uses `urllib.parse.urlparse` to split the URL into its components. Then I use `splitext` to split the archive extension from the path in the URL. 

### How did you test it?
Manual test with the To Reproduce instruction from #5198


### Checklist

- [X] I have read the [contributors guidelines](https://github.com/deepset-ai/haystack/blob/main/CONTRIBUTING.md) and the [code of conduct](https://github.com/deepset-ai/haystack/blob/main/code_of_conduct.txt)
- [X] I have updated the related issue with new insights and changes
- [x] I added unit tests and updated the docstrings
- [X] I've used one of the [conventional commit types](https://www.conventionalcommits.org/en/v1.0.0/) for my PR title: `fix:`, `feat:`, `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`.
- [x] I documented my code
- [X] I ran [pre-commit hooks](https://github.com/deepset-ai/haystack/blob/main/CONTRIBUTING.md#installation) and fixed any issue

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/deepset-ai/haystack?pullRequest=5199) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/deepset-ai/haystack?pullRequest=5199) before we can accept your contribution.<br/><hr/>**Malte Reimann** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/deepset-ai/haystack?pullRequest=5199) it.</sub>

anakin87 commented:
Hello @malte-aws and thanks for the PR, that is useful and appreciated! :+1: 

Some suggestions:
- the CI is failing because of black and pylint. If you install [precommit](https://github.com/deepset-ai/haystack/blob/main/CONTRIBUTING.md#installation), it can automatically solve these small style problems.
- I would extract the logic and create a function `get_filename_extension_from_url` (just an example)
- Now you should add some unit tests for this function in https://github.com/deepset-ai/haystack/blob/main/test/others/test_utils.py (testing for example one URL ending with an extension and another similar to the URL given at https://github.com/deepset-ai/haystack/issues/5198#issue-1771131219)

Let us know if you need any help and thanks again!

malte-aws commented:
Hi @anakin87,

thank you for the feedback.

I addressed the improvements by
1. Running pre-commit
2. Extracting the code that gets the filename and extension into a separate function
3. Added unit test covering the code contribution.

Kind regards
Malte

anakin87 commented on haystack/utils/import_utils.py:
```diff
@@ -93,6 +95,20 @@ def load_documents_from_hf_datasets(dataset_name: str, split: Optional[str] = "t
+    root, extension = splitext(parsed.path)
+    archive_extension = extension[1:]
+    file_name = unquote(basename(root[1:]))
+    return (file_name, archive_extension)
```
you can skip the parentheses
```suggestion
    return file_name, archive_extension
```

anakin87 commented on test/others/test_utils.py:
```diff
@@ -168,6 +168,20 @@ def test_convert_files_to_docs(samples_path):
     assert documents and len(documents) > 0
 
 
+def test_get_filename_extension_from_url_zip():
```
Please mark this test as a unit test.
We are reorganizing the tests, so in some cases, the marker is missing but is needed to execute them.

```suggestion
@pytest.mark.unit
def test_get_filename_extension_from_url_zip():
```

malte-aws replied:
[ee1f6c8](https://github.com/deepset-ai/haystack/pull/5199/commits/ee1f6c8ed51411a43d6c8dabd66183f8d2eda670) add marker

anakin87 commented on test/others/test_utils.py:
```diff
@@ -168,6 +168,20 @@ def test_convert_files_to_docs(samples_path):
+    assert file_name == "filename"
+
+
+def test_get_filename_extension_from_url_xz():
```
Same as above.

```suggestion
@pytest.mark.unit
def test_get_filename_extension_from_url_xz():
```

malte-aws replied:
Resolved by #[ee1f6c8](https://github.com/deepset-ai/haystack/pull/5199/commits/ee1f6c8ed51411a43d6c8dabd66183f8d2eda670)

anakin87 reviewed (commented):
This PR is in a very good state.

It would be great if you could add a test for a "normal" URL, such as http://www.mysite.com/resources/myfile.zip

anakin87 reviewed (approved):
Looks good to me!

Thank you once again for your first PR on Haystack :partying_face:
